boards/posix/native_posix/fuse_flash.c

```diff
@@ -0,0 +1,531 @@
+		} else {
+			stat.st_mode = S_IFREG | S_IRW_FILE;
+			stat.st_size = entry.size;
+		}
+
+		if (filler(buf, entry.name, &stat, 0)) {
+			break;
+		}
+
+	} while (1);
+
+	fs_closedir(&dir);
+
+	return err;
+
+}
+
+static int fuse_flash_create(const char *path, mode_t mode,
+			     struct fuse_file_info *fi)
+{
+	int err;
+	size_t handle;
+
+	ARG_UNUSED(mode);
+
+	if (is_mount_point(path)) {
+		return -ENOENT;
+	}
+
+	handle = get_new_file_handle();
+	if (handle == INVALID_FILE_HANDLE) {
+		return -ENOMEM;
+	}
+
+	fi->fh = handle;
+
+	err = fs_open(&files[handle], path);
+	if (err != 0) {
+		release_file_handle(handle);
+		fi->fh = INVALID_FILE_HANDLE;
+		return err;
+	}
+
+	return 0;
+}
+
+static int fuse_flash_open(const char *path, struct fuse_file_info *fi)
+{
+	return fuse_flash_create(path, 0, fi);
+}
+
+static int fuse_flash_release(const char *path, struct fuse_file_info *fi)
+{
+	ARG_UNUSED(path);
+
+	if (fi->fh == INVALID_FILE_HANDLE) {
+		return -EINVAL;
+	}
+
+	fs_close(&files[fi->fh]);
+
+	release_file_handle(fi->fh);
+
+	return 0;
+}
+
+static int fuse_flash_read(const char *path, char *buf, size_t size,
+		off_t off, struct fuse_file_info *fi)
+{
+	int err;
+
+	ARG_UNUSED(path);
+
+	if (fi->fh == INVALID_FILE_HANDLE) {
+		return -EINVAL;
+	}
+
+	err = fs_seek(&files[fi->fh], off, FS_SEEK_SET);
+	if (err != 0) {
+		return err;
+	}
+
+	err = fs_read(&files[fi->fh], buf, size);
+
+	return err;
+}
+
+static int fuse_flash_write(const char *path, const char *buf, size_t size,
+		off_t off, struct fuse_file_info *fi)
+{
+	int err;
+
+	ARG_UNUSED(path);
+
+	if (fi->fh == INVALID_FILE_HANDLE) {
+		return -EINVAL;
+	}
+
+	err = fs_seek(&files[fi->fh], off, FS_SEEK_SET);
+	if (err != 0) {
+		return err;
+	}
+
+	err = fs_write(&files[fi->fh], buf, size);
+
+	return err;
+}
+
+static int fuse_flash_ftruncate(const char *path, off_t size,
+				struct fuse_file_info *fi)
+{
+	int err;
+
+	ARG_UNUSED(path);
+
+	if (fi->fh == INVALID_FILE_HANDLE) {
+		return -EINVAL;
+	}
+
+	err = fs_truncate(&files[fi->fh], size);
+
+	return err;
+}
+
+static int fuse_flash_truncate(const char *path, off_t size)
+{
+	int err;
+	static struct fs_file_t file;
+
+	err = fs_open(&file, path);
+	if (err != 0) {
+		return err;
+	}
+
+	err = fs_truncate(&file, size);
+	if (err != 0) {
+		fs_close(&file);
+		return err;
+	}
+
+	err = fs_close(&file);
+
+	return err;
+}
+
+static int fuse_flash_mkdir(const char *path, mode_t mode)
+{
+	ARG_UNUSED(mode);
+
+	return fs_mkdir(path);
+}
+
+static int fuse_flash_rmdir(const char *path)
+{
+	return fs_unlink(path);
+}
+
+static int fuse_flash_unlink(const char *path)
+{
+	return fs_unlink(path);
+}
+
+static int fuse_flash_statfs(const char *path, struct statvfs *buf)
+{
+	ARG_UNUSED(path);
+	ARG_UNUSED(buf);
+	return 0;
+}
+
+static int fuse_flash_utimens(const char *path, const struct timespec tv[2])
+{
+	/* dummy */
+	ARG_UNUSED(path);
+	ARG_UNUSED(tv);
+
+	return 0;
+}
+
+
+static struct fuse_operations fuse_flash_oper = {
```

Comment:
I have no experience with the FUSE libray.. but seeing all this code I really wonder how stable its API is..

Reply:
I do understand your concern but actually there is a lot of backwards compatibility in the library.
You may have notice the line ```#define FUSE_USE_VERSION 26``` this actually needs to be declared before including the fuse header file and specifies the API version used.